server/src/lib/serial.rs

```diff
@@ -172,15 +172,19 @@ impl<Ctx: 'static, Device: Sink<Ctx> + Source<Ctx>> Serial<Ctx, Device> {
     }
 
     /// Returns a future that completes when the read buffer becomes full.
-    pub async fn read_buffer_full(&mut self) {
-        while !self.source_driver.lock().unwrap().source_full {
-            // By relaying on the "notify_one" permit mechanism, we are safe
-            // from a TOCTTOU race condition where the buffer becomes full
-            // after we check "source_full" but before we await the notified
-            // signal.
-            //
-            // In this case, the "notified()" call will return immediately.
-            self.source_full_signal.notified().await;
+    pub fn read_buffer_full(&self) -> impl futures::Future<Output = ()> {
```

Comment:
Because the anonymouse future the `async fn` version returns is linked to the lifetime of `self` here. That ends up tripping "cannot borrow serial because it's already borrowed type errors". The explicit returning-a-future version here lets us return a Future that isn't tied to self's lifetime since the underlying fields in self we use here are already behind an Arc that we can clone.